model.py

```diff
@@ -9,43 +9,98 @@
+          * a tensor of shape [seq_len, batch_size] containing positions."""
+        # TODO(neil): rewrite so that we don't need seq_len leading dimension
+        maskedEmbedding = textEmbedding * text_mask.unsqueeze(-1)
+        seq_len = query.shape[0]
+        batch_size = query.shape[1]
+        if position is None:
+            position = torch.zeros((seq_len, batch_size), device=query.device)
+        query_result = torch.exp(self.linear(query))
+        assert query_result.shape == (seq_len, batch_size, 2)
+        assert textEmbedding.shape[1] == batch_size
+        mean, std = query_result.unbind(dim=-1)
+        if self.use_position:
+            mean = position + mean.cumsum(dim=0)/20
+        weights = torch.zeros((seq_len, textEmbedding.shape[0], batch_size, 1), device=query.device)
+        # TODO(neil): Why these extra dimensions with value?
+        indices = torch.arange(0, textEmbedding.shape[0], device=query.device).view(1, textEmbedding.shape[0], 1, 1)
+        weights = torch.exp(-std.view(seq_len, 1, batch_size, 1) * (mean.view(seq_len, 1, batch_size, 1) - indices) ** 2)
+        weightedSum = maskedEmbedding * weights
+        newContext = weightedSum.sum(dim=1)
+
+        return newContext, mean
+
 class Scribe(nn.Module):
     @argbind.bind(without_prefix=True)
-    def __init__(self, input_size=3, hidden_size=20, num_layers=1, output_size=3):
+    # TODO(neil): Update num_embeddings from data
+    def __init__(self, input_size=3, hidden_size=20, num_layers=1, output_size=3, num_embeddings=100, use_position=1):
         super(Scribe, self).__init__()
-        self.rnn = nn.GRU(input_size=input_size, hidden_size=hidden_size, num_layers=num_layers)
+        self.embedding_dim = 10
+        self.embedding = nn.Embedding(num_embeddings=num_embeddings, embedding_dim=self.embedding_dim)
+        self.hidden_size = hidden_size
+        self.rnn1 = nn.GRU(input_size=input_size, hidden_size=hidden_size, num_layers=1)
+        self.rnn2 = nn.GRU(input_size=hidden_size+self.embedding_dim, hidden_size=hidden_size, num_layers=max(1, num_layers-1))
         self.linear = nn.Linear(in_features=hidden_size, out_features=output_size)
+        self.attention = GaussianAttention(query_size=hidden_size,use_position=use_position == 1)
+        self.textConverter = data.Data()
+
 
-    def forward(self, strokes):
+    def forward(self, strokes, texts, textsmask):
         hidden = None
-        output, hidden = self.step(strokes, hidden)
+        textEmbedding = self.embedding(texts)
+        output, hidden = self.step(strokes, textEmbedding, textsmask, hidden)
         return output
 
-    def step(self, x, hidden):
-        x, hidden = self.rnn(x, hidden)
-        output = self.linear(x)
-        return output, hidden
+    def step(self, x, textEmbedding, textsmask, hidden):
+        if hidden is None:
+            hidden1 = None
+            positions = None
+            hidden3 = None
```

Comment:
Not a huge thing but I'd usually name this `hidden2` for consistency. Otherwise, when I read the code I think... 'Where's hidden2?'

Reply:
Renamed.  Good point.